### PR TITLE
Fix/tr 6156/open links in a new tab text reader interaction

### PR DESCRIPTION
**Pull request by Nicholass**
https://oat-sa.atlassian.net/browse/TR-6156

Now all HTML anchor tags inside the text reader interaction  are getting `target` set to `_blank`, even if it was not present or was set to different value. We want all the links to open externally

### CODE REVIEW

File: views/js/pciCreator/dev/textReaderInteraction/runtime/js/renderer.js [hunk 1]
```diff
@@ -27,8 +27,25 @@ define(
...
         };
         
-        return function (options) {
-            var self = this;
+        /**
+         * Add all html anchors a "_blank" target by default
+         *
+         * @param {String} html - the html to parse
+         * @param {Object} renderer
+         * @returns {String} the html without updated URLs
```
Comment (oatymart) [inline views/js/pciCreator/dev/textReaderInteraction/runtime/js/renderer.js:35]:
```suggestion
         * @returns {String} the html with fixed anchor targets
```

File: views/js/pciCreator/ims/textReaderInteraction/runtime/js/renderer.js [hunk 2]
```diff
@@ -173,7 +173,17 @@ define(
                                 image.setAttribute('src', content);
                             }
                         });
-                        return element.outerHTML || element.textContent;
+
+                        anchors = selectorContainer.querySelectorAll('a');
```
Comment (oatymart) [inline views/js/pciCreator/ims/textReaderInteraction/runtime/js/renderer.js:177]: This one needs `var anchors;`, above.

File: views/js/pciCreator/dev/textReaderInteraction/runtime/js/renderer.js [hunk 3]
```diff
@@ -27,8 +27,25 @@ define(
...
+
+            return html.replace(/(<a\s+[^>]*)(?=\s*>)/ig, function(substr, $1) {
+                // Remove any existing target attribute and replace with target="_blank"
+                var modifiedTag = $1.replace(/\btarget\s*=\s*["'][^"']*["']/ig, '').trim();
+                return modifiedTag + ' target="_blank"';
+            });
+        };
+      
+      return function (options) {
+        var self = this;
```
Comment (oatymart) [inline views/js/pciCreator/dev/textReaderInteraction/runtime/js/renderer.js:48]: Indentation

**Review by oatymart — CHANGES_REQUESTED** [review]
The code has a couple of mistakes but the principle is ok and I verified that after fixing, it will do its job well.

There are also a few extra things needed to complete the PR:
- Bump the versions in the 2 manifests: `dev/textReaderInteraction/pciCreator.json` and `ims/textReaderInteraction/imsPciCreator.json`
- When the code is approved, the minified files need to be added to the PR. The command is (in `tao/views/build`): `npx grunt portableelement -e=pciSamples -i=textReaderInteraction`
- A migration should be created to install the new version of the IMS PCI (1.2.2). See [this example](https://github.com/oat-sa/extension-pcisample/commit/beb36b2faabadf2806914d0999712f288ee7054d#diff-e7ff910511f98a5804547ead34deca30e9418cf4bc0b4cc35d21051f6bcc03c8) and be aware that there is a CLI for generating, running and rolling back the migrations ([documented here](https://oat-sa.atlassian.net/wiki/spaces/~456852655/pages/1408665016/Migrations+Example))

**Comment by oatymart** [comment]
Please add PR description too. It's a must-have.

File: views/js/pciCreator/dev/textReaderInteraction/runtime/js/renderer.js [hunk 4]
```diff
@@ -177,6 +195,11 @@ define(
                             markup,
                             self.options.interaction.renderer
                         );
+
+                        fixedMarkup = fixAnchorTargets(
+                            markup,
+                            self.options.interaction.renderer
+                        );
```
[the remark this answers is not in the document]
Comment (Nicholass) [reply views/js/pciCreator/dev/textReaderInteraction/runtime/js/renderer.js]: now HTML fixes are performed consecutively

File: views/js/pciCreator/dev/textReaderInteraction/runtime/js/renderer.js (hunk 1 above)
Comment (Nicholass) [reply views/js/pciCreator/dev/textReaderInteraction/runtime/js/renderer.js]: fixed docstring

**Review by wazelin — CHANGES_REQUESTED** [review]
In order for TAO 3.x to register the updated PCI runtime and creator code, the PCI must be registered via a migration.
See https://github.com/oat-sa/extension-pcisample/blob/beb36b2faabadf2806914d0999712f288ee7054d/migrations/Version202406120835514106_pciSamples.php as an example.

Such a migration can be added by running [this command](https://github.com/oat-sa/tao-core/blob/51d0c2e9fd1e273b250b30483f00fc99a292787f/scripts/tools/Migrations.php#L76), i.e. `docker exec nextgen_tao php index.php '\oat\tao\scripts\tools\Migrations' -c generate -e pciSamples`.

Be sure to _install_ the extension first if it hadn't been already: `docker exec nextgen_tao php tao/scripts/installExtension.php pciSamples`.

File: migrations/Version202407021254084106_pciSamples.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+  declare(strict_types=1);
+
+  namespace oat\pciSamples\migrations;
+
+  use Doctrine\DBAL\Schema\Schema;
+  use oat\tao\scripts\tools\migrations\AbstractMigration;
```
Comment (oatymart) [inline migrations/Version202407021254084106_pciSamples.php:8]:
```suggestion
  use Doctrine\DBAL\Schema\Schema;
  use Doctrine\Migrations\Exception\IrreversibleMigration;
  use oat\pciSamples\scripts\install\RegisterPciTextReaderIMS;
  use oat\tao\scripts\tools\migrations\AbstractMigration;
```
You need to `use` all classes you reference in the script, or the migration will error.

File: migrations/Version202407021254084106_pciSamples.php
Comment (oatymart) [inline migrations/Version202407021254084106_pciSamples.php:1]: Please check the file indentation against previous PHP files. I think it is all indented by 2 spaces too much.

**Review by oatymart — CHANGES_REQUESTED** [review]
Please test your migration:
Up:
`docker exec nextgen_tao php index.php '\oat\tao\scripts\tools\Migrations' -c execute -v 'oat\pciSamples\migrations\Version202407021254084106_pciSamples'`
Down:
`docker exec nextgen_tao php index.php '\oat\tao\scripts\tools\Migrations' -c rollback -v 'oat\pciSamples\migrations\Version202407021254084106_pciSamples'`

Please also generate and commit the JS bundles for the 2 PCIs:
```
cd tao/views/build
npm ci
npx grunt portableelement -e=pciSamples -i=textReaderInteraction
```

**Review by wazelin — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`

**Review by oatymart — APPROVED** [review]
Nice one, the extension seems fully working.

**Pull request merged**